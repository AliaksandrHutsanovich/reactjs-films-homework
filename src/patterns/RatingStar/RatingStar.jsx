import React, { useMemo, memo } from 'react';
import PropTypes from 'prop-types';
import { List as AntdList, Row, Col } from 'antd';
import { List, Text } from '../../components';

import '!style-loader!css-loader!sass-loader!antd/dist/antd.css'; //eslint-disable-line

import styles from './RatingStar.scss';

const { Item } = AntdList;

export const renderItem = (item) => (
  <Item className={styles.item}>
    {item.star}
  </Item>
);

const RatingStar = ({
  rating,
}) => {
  const frontStarsArray = useMemo(() => Array(5).fill(
    <Text type="secondary" size={26}>&#9733;</Text>,
  ), []);

  const backStarsArray = useMemo(() => Array(5).fill(
    <Text type="white" size={26}>&#9733;</Text>,
  ), []);

  const backList = useMemo(
    () => Array.from(backStarsArray, (el, i) => ({ id: i, star: el })),
    [backStarsArray],
  );
  const frontList = useMemo(
    () => Array.from(frontStarsArray, (el, i) => ({ id: i, star: el })),
    [frontStarsArray],
  );

  return (
    <Row justify="start" gutter={10}>
      <Col className={styles.patternsWrapper}>
        <div className={styles.backPattern}>
          <List
            dataSource={backList}
            grid={{ gutter: 6, column: 5 }}
            renderItem={renderItem}
          />
        </div>
        <div
          style={{
            width: `${(rating / 5) * 100}%`,
          }}
          className={styles.frontPatternWrapper}
        >
          <div className={styles.frontPattern}>
            <List
              dataSource={frontList}
              grid={{ gutter: 6, column: 5 }}
              renderItem={renderItem}
            />
          </div>
        </div>
      </Col>
      <Col>
        <Text type="white_with_background" size={12}>
          {rating}
        </Text>
      </Col>
    </Row>
  );
};

// &#9733; this is a filled star, it as more apropriate

RatingStar.propTypes = {
  rating: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]).isRequired,
};

export default memo(RatingStar);
