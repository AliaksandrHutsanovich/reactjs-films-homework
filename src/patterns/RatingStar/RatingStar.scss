.patternsWrapper {
  padding-left: 0px !important;
  padding-right: 0px !important;
}

.backPattern {
  position: relative;
  width: 120px;
}

.frontPatternWrapper {
  overflow: hidden;
  position: absolute;
  top: 0px;
}

.frontPattern {
  width: 120px;
}

.item {
  padding: 0 0;
  margin-top: -10px;
  margin-bottom: 0px !important;
}
