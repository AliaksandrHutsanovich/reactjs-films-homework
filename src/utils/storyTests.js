const storyTests = (stories, renderer) => {
  Object.keys(stories).forEach((storyKey) => {
    if (typeof stories[storyKey] !== 'function') {
      return;
    }
    it(`${storyKey} should be rendered properly`, () => {
      renderer.render(stories[storyKey]());
      const result = renderer.getRenderOutput();

      expect(result).toMatchSnapshot();
    });
  });
};

export default storyTests;
