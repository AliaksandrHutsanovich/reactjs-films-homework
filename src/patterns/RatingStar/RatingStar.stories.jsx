import React from 'react';
import RatingStar from './RatingStar';

export default {
  component: RatingStar,
  title: 'patterns/RatingStar',
  decorators: [
    (story) => (
      <div
        style={{
          padding: '20px',
          width: '220px',
          height: '90px',
          backgroundColor: 'red',
        }}
      >
        {story()}
      </div>
    ),
  ],
};

export const ratingStarExample = () => (
  <RatingStar rating="4.6" />
);
