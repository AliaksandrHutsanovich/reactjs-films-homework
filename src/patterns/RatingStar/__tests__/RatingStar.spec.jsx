import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import storyTests from '../../../utils/storyTests';
import * as stories from '../RatingStar.stories';
import { renderItem } from '../RatingStar';

describe('<RatingStar />', () => {
  const renderer = new ShallowRenderer();

  storyTests(stories, renderer);

  it('star shold be rendered properly', () => {
    const Star = ({ item }) => renderItem(item);
    renderer.render(<Star item={{ id: 0, star: '&#9733;' }} />);
    const result = renderer.getRenderOutput();

    expect(result).toMatchSnapshot();
  });
});
