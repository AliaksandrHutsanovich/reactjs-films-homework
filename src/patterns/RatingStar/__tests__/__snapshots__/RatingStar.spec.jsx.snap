// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RatingStar /> ratingStarExample should be rendered properly 1`] = `
<ForwardRef
  gutter={10}
  justify="start"
>
  <ForwardRef
    className="patternsWrapper"
  >
    <div
      className="backPattern"
    >
      <UNDEFINED
        dataSource={
          Array [
            Object {
              "id": 0,
              "star": <UNDEFINED
                size={26}
                type="white"
              >
                ★
              </UNDEFINED>,
            },
            Object {
              "id": 1,
              "star": <UNDEFINED
                size={26}
                type="white"
              >
                ★
              </UNDEFINED>,
            },
            Object {
              "id": 2,
              "star": <UNDEFINED
                size={26}
                type="white"
              >
                ★
              </UNDEFINED>,
            },
            Object {
              "id": 3,
              "star": <UNDEFINED
                size={26}
                type="white"
              >
                ★
              </UNDEFINED>,
            },
            Object {
              "id": 4,
              "star": <UNDEFINED
                size={26}
                type="white"
              >
                ★
              </UNDEFINED>,
            },
          ]
        }
        grid={
          Object {
            "column": 5,
            "gutter": 6,
          }
        }
        renderItem={[Function]}
      />
    </div>
    <div
      className="frontPatternWrapper"
      style={
        Object {
          "width": "92%",
        }
      }
    >
      <div
        className="frontPattern"
      >
        <UNDEFINED
          dataSource={
            Array [
              Object {
                "id": 0,
                "star": <UNDEFINED
                  size={26}
                  type="secondary"
                >
                  ★
                </UNDEFINED>,
              },
              Object {
                "id": 1,
                "star": <UNDEFINED
                  size={26}
                  type="secondary"
                >
                  ★
                </UNDEFINED>,
              },
              Object {
                "id": 2,
                "star": <UNDEFINED
                  size={26}
                  type="secondary"
                >
                  ★
                </UNDEFINED>,
              },
              Object {
                "id": 3,
                "star": <UNDEFINED
                  size={26}
                  type="secondary"
                >
                  ★
                </UNDEFINED>,
              },
              Object {
                "id": 4,
                "star": <UNDEFINED
                  size={26}
                  type="secondary"
                >
                  ★
                </UNDEFINED>,
              },
            ]
          }
          grid={
            Object {
              "column": 5,
              "gutter": 6,
            }
          }
          renderItem={[Function]}
        />
      </div>
    </div>
  </ForwardRef>
  <ForwardRef>
    <UNDEFINED
      size={12}
      type="white_with_background"
    >
      4.6
    </UNDEFINED>
  </ForwardRef>
</ForwardRef>
`;

exports[`<RatingStar /> star shold be rendered properly 1`] = `
<Item
  className="item"
>
  &#9733;
</Item>
`;
